tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,257 @@
+        await screenshot(f"http://{unit_ip}", screenshot_dir / f"wordpress-before-{idx}.png")
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_upgrade(
+    ops_test: pytest_operator.plugin.OpsTest,
+    application_name,
+    gen_upgrade_test_charm_config,
+    num_units,
+):
+    """
+    arrange: an old version of the WordPress is deployed.
+    act: build WordPress charm from source and upgrade the WordPress charm.
+    assert: all operations finished without error.
+    """
+
+    charm = await ops_test.build_charm(".")
+    await ops_test.model.remove_application(application_name)
+
+    def wordpress_removed():
+        status = subprocess.check_output(
+            ["juju", "status", "-m", ops_test.model_name, "--format", "json"]
+        )
+        return application_name not in json.loads(status)["applications"]
+
+    await ops_test.model.block_until(wordpress_removed, wait_period=5, timeout=600)
+    await ops_test.model.deploy(
+        str(charm),
+        resources={"wordpress-image": "localhost:32000/wordpress:test"},
+        application_name=application_name,
+        series="jammy",
+        num_units=num_units,
+        config=gen_upgrade_test_charm_config(),
+    )
+    await ops_test.model.wait_for_idle(status=ops.model.ActiveStatus.name)  # type: ignore
+
+
+@pytest.mark.abort_on_fail
+async def test_wordpress_after_upgrade(unit_ip_list, application_name, screenshot_dir):
+    """
+    arrange: the WordPress charm has been upgraded.
+    act: browser the WordPress website powered by the new charm.
+    assert: the website should have the same content as the old one.
+    """
+
+    def check_images(html):
+        for url in re.findall('<img[^>]+src="([^"]+)"[^>]*>', html):
+            logger.info("check image %s", url)
+            image_response = requests.get(url, timeout=10)
+            assert (
+                image_response.status_code == 200
+            ), f"access image {url} should return status 200"
+            try:
+                PIL.Image.open(io.BytesIO(image_response.content), formats=(url.split(".")[-1],))
+            except PIL.UnidentifiedImageError:
+                raise AssertionError(f"access image {url} should return a valid image file")
+
+    for idx, unit_ip in enumerate(unit_ip_list):
+        await screenshot(f"http://{unit_ip}", screenshot_dir / f"wordpress-after-{idx}.png")
+        after_image = PIL.Image.open(screenshot_dir / f"wordpress-after-{idx}.png")
+        before_image = PIL.Image.open(screenshot_dir / f"wordpress-before-{idx}.png")
+        comparison_width = before_image.width + after_image.width
+        comparison_height = max(before_image.height, after_image.height)
+        comparison_image = PIL.Image.new("RGB", (comparison_width, comparison_height))
+        comparison_image.paste(before_image, (0, 0))
+        comparison_image.paste(after_image, (before_image.width, 0))
+        comparison_image.save(screenshot_dir / f"wordpress-comparison-{idx}.png")
```

Comment:
If I understand well you creating an image that is an overlay of the 2 screenshots. Is it meant to be manually checked ? How do we ensure this is even checked ?

Reply:
This is a side by side comparison, and for now it's manually checked. It's difficult for machine to check it since the charm upgrade may cause the theme upgrade, which will alter the appearance of the blog and that's completely normal. I have some plan in the future that post the comparison to the PR page as a comment, so the changes can be seen by code author and reviewers.

Reply:
Would it be worth it to include a comment explaining this in the code?

Reply:
Added, thanks!